File: vol_sfera.cxx
#include <iostream>
#define _USE_MATH_DEFINES
#include <cmath>
#include <vector>
#include <random>
#include <chrono>

using namespace std;

int main(){
    //typedef std::chrono::high_resolution_clock myclock;
    //myclock::time_point beginning = myclock::now();
    //myclock::duration d = myclock::now() - beginning;
    //unsigned seed2 = d.count();
    unsigned seed2 = chrono::system_clock::now().time_since_epoch().count();
    int m, N=10000000;
    double vol_t, vol, omega, media=0, somma=0, p, r, temp, sigma, var;
    vector <double> x_rand;
    vector <double> func;
    default_random_engine rd(seed2);
    
    //mt19937 gen(rd());
    uniform_real_distribution <double> dis(-1,1);
    cout<<"inserire numero di dimensioni: "<<endl;
    cin>>m;

    //theoretical volume of the sphere
    vol_t=pow(M_PI,m/2)/tgamma((m/2)+1);
    cout<<"volume teorico: "<<vol_t<<endl;
    //integration domain
    omega=pow(2,m);

    for (int i=0; i<N; i++){
        for (int j=0; j<m; j++){ //this is to generate m values for the components
            r=dis(rd);
            x_rand.push_back(r);
            temp+=pow(r,2);
        }
        if(temp<=1){
            media+=1;
            func.push_back(1);
        }
        else{
            func.push_back(0);
        }
        temp=0;
    }
    media=media/N;
    vol=omega*media;

    cout<<"Il volume calcolato e': "<<vol;

    for(int k=0; k<func.size(); k++){
        somma+=pow(func.at(k)-media,2);
    }
    var=somma/N;
    sigma=sqrt(var/N);
    cout<<" +/- "<<sigma<<endl;




    return 0;
}

File: vol_sfera.py
import scipy as sc
import numpy as np
import math 

#numeri random tra -1 e 1
def random(m):
    x=np.random.uniform(-1,1,(m))
    return x

def main():
    N=1000000
    M=int(input("Inserire dimensione dello spazio:\n"))
    omega=pow(2,M)
    gamma=sc.special.gamma((M/2)+1)
    Vol_t=pow(math.pi, M/2)/gamma
    print("Valore di riferimento:")
    print(Vol_t)

    media=0
    funz=[]
    for i in range(N):
        x=random(M)
        p=np.dot(x,x)
        if(p<=1):
            media+=1
            funz.append(1)
        else:
            funz.append(0)
    media=media/N
    Vol=omega*media
    print("Valore stimato:")
    print(Vol)
    
    somma=0
    somma2=0
    for j in range(len(funz)):
        somma+=pow(funz[j]-media,2)
        somma2+=pow(funz[j],2)
        
    #somma2=somma2/N 
    #var=somma2-media**2
    var=somma/N
    
    sigma=np.sqrt(var/N)
    print(sigma)
    #print(1/np.sqrt(N))

if __name__=='__main__':
    main()